### DISCUSSION TITLE
Feature: Support package.json exports fields

### DISCUSSION THREAD

**Discussion by flean** [meteor/meteor#11727] [Feature Requests]
Right now I'm stuck with version 6 of swiper and would like to upgrade to 7 but it requires ESM support. 

Is there a way for us to use ESM packages in meteor?

[SwiperJS](https://swiperjs.com/migration-guide)

>  the title was firstly named: ESM packages support

**Comment by StorytellerCZ** [comment]
I don't follow, are you getting an error when you are trying to import from Swipper 7?

For example:
```js
import Swiper from 'swiper/swiper-bundle.esm.js';
```
or
```js
import Swiper from 'swiper';
```

**Comment by flean** [comment]
```
// Import Swiper React components
import { Swiper, SwiperSlide } from 'swiper/react';

// Import Swiper styles
import 'swiper/css';

```

**Comment by StorytellerCZ** [comment]
Shouldn't it be?
```js
import Swiper, { SwiperSlide } from 'swiper/react';
```

**Comment by flean** [comment]
No, that comes directly from their docs and the react component inherits swiper core. At least thats my understanding.

**Comment by StorytellerCZ** [comment]
I still don't understand what the problem here is.

**Comment by flean** [comment]
It can’t find the module, app breaks. I have to use version 6 for it to work. It looks like meteor can’t use ESM packages. I’ll setup a meteor app only loading that component.

**Comment by flean** [comment]
here is the [repo](https://github.com/flean/simple-swiper-example) with version 7

Server
```
Unable to resolve some modules:

  "swiper/css" in /simple-swiper-example/imports/ui/App.jsx (web.browser)

If you notice problems related to these missing modules, consider running:

  meteor npm install --save swiper
```
 Client Side
```
[Log] HMR: connected (hot-module-replacement.js, line 330)
[Error] Error: Cannot find module 'swiper/react'
	(anonymous function) (modules-runtime-hot.js:253)
	moduleLink (modules.js:300)
	module (app.js:55)
	fileEvaluate (modules-runtime-hot.js:388)
	require (modules-runtime-hot.js:270)
	moduleLink (modules.js:307)
	module (app.js:218)
	fileEvaluate (modules-runtime-hot.js:388)
	require (modules-runtime-hot.js:270)
	Global Code (app.js:240)
```

**Comment by StorytellerCZ** [comment]
OK, I see now!

Yeah, that is indeed not working. Right now workaround would be to import from the files directly like this:
```js
import 'swiper/swiper.min.css';
```

But that is only a stop gap solution.

@zodern @renanccastro @filipenevola 
The issue here is that in `package.json` the exports are defined under `exports` prop, in this case like this:
```json
"exports": {
    ".": "./swiper.esm.js",
    "./core": "./swiper.esm.js",
    "./bundle": "./swiper-bundle.esm.js",
    "./css": "./swiper.min.css",
    "./css/bundle": "./swiper-bundle.min.css",
    "./css/a11y": "./modules/a11y/a11y.min.css",
    "./css/autoplay": "./modules/autoplay/autoplay.min.css",
...
```

**Comment by trusktr** [comment]
Even TypeScript barely [_just_ added](https://github.com/microsoft/TypeScript/issues/33079#issuecomment-926936955) support for recognizing package.json `exports` fields. Meteor's turn now.

The only solution right now is to import the files directly until Meteor adds support for it as well.

Keep one thing in mind about this workaround though: importing files directly will work in Meteor, but will fail in Node apps, or in Webpack which also follows the latest Node ESM spec. If you have some code in package A, for example, and it imports files directly from some package B, this will work in Meteor, but it will throw an error in Node or Webpack apps.

**Comment by trusktr** [comment]
This is a duplicate of https://github.com/meteor/meteor/issues/10906

**Comment by StorytellerCZ** [comment]
Some notes from me digging around this:
Babel plans to become ESM package in Babel 8: https://babeljs.io/docs/en/roadmap#babel-8
There is this esm package which should make things work: https://www.npmjs.com/package/esm
But Node should supporting this already: https://nodejs.org/docs/latest-v14.x/api/esm.html

**Comment by make-github-pseudonymous-again** [comment]
Is this something that only needs an upstream (upcoming) update of TypeScript?

**Reply by trusktr** [reply]
For TypeScript support specifically, yes. For now, importing packages with an `exports` field in package.json won't work (it might, if you get lucky, but it depend on each package). https://github.com/microsoft/TypeScript/issues/33079

For Meteor in general, no, because package.json `exports` field and Node ESM module resolution doesn't have to do with TypeScript (TypeScript doesn't support the features yet, but once it does, Meteor still needs to build file correctly, independently of TypeScript).

**Comment by StorytellerCZ** [comment]
I don't think that will be it. I think the problem here for me is to pin the issue down. Meteor already supports what is commonly known as JavaScript modules.  What we are talking about here is a new feature which allows developers to specify their exports under export param in `package.json` which we can't process here. Given that we have special handling for things like inline imports, maybe we need to update other parts as well. Any thoughts @zodern @renanccastro ?

**Reply by zodern** [reply]
It seems this mostly affects how modules are resolved. In Meteor this would be implemented [here in install](https://github.com/benjamn/install/blob/916924060877a846956158542165e90e457bb220/install.js#L498-L511) and [here](https://github.com/meteor/meteor/blob/4046800590ae5ffc409e1466f44dbf4641c9c736/tools/isobuild/resolver.ts#L130) in the Meteor tool. [This](https://github.com/jkrems/proposal-pkg-exports) seems to be the current specification.

**Comment by StorytellerCZ** [comment]
Here is the relevant docs from Node.js: https://nodejs.org/api/packages.html#packages_exports
TypeScript is implementing this in 4.5: https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-beta/#packagejson-exports-imports-and-self-referencing

**Reply by trusktr** [reply]
Nice.

Regardless of TypeScript though, this needs to be implemented in Meteor's loader (`meteorInstall` and related tooling).

**Comment by edemaine** [comment]
I just ran into this issue, specifically the desire for *export conditions* (see [this Node documentation](https://nodejs.org/api/packages.html#community-conditions-definitions)). I'm using the `solid-js` package, which offers a regular `main` entry point, but also uses `exports` to define specific exports when in "development" mode with additional debugging code.

So when implementing support for `exports`, it'd be really sweet for `meteor run` to enable the "development" condition, while `meteor build` would not.

**Comment by StorytellerCZ** [comment]
Found a related issue in Reify with comment from @benjamn : https://github.com/benjamn/reify/issues/329

**Comment by trusktr** [comment]
> There is this esm package which should make things work: https://www.npmjs.com/package/esm

The `esm` package is out of date and does not support Node ESM spec, so it will fail to import various Node ESM packages from node_modules.

> Node should supporting this already: https://nodejs.org/docs/latest-v14.x/api/esm.html

Node supports it, but code in Meteor that passes through Meteor's build system `ecmascript` by @benjamn won't work with Node ESM format. Meteor needs to

- update `ecmascript` to support it (reify issue in the previous comment), and
- make it easy to not use `ecmascript` on the server so that native Node ESM can be used instead

No matter what, if we're bundling code for the client, then we need to support Node ESM format to be compatible with packages that are moving forward to Node ESM.

Package authors should avoid the `exports` field of package.json so as to be compatible with both TypeScript and Meteor.

**Reply by edemaine** [reply]
Using native ESM on server would be nice, but given that Reify/Meteor `modules` already supports `import` syntax well, an alternative approach is to just emulate the `exports` feature. I don't *think* it would be hard, but I don't know what isometric assumptions Reify makes. Meteor is already not isometric between client and server (e.g. legacy vs not builds) so I feel like it should be possible.

**Reply by eric-burel** [reply]
I don't get the "avoid exports" bit. I've noticed that Apollo does avoid it, however exports map seem to be a core feature of npm. I've written my own packages using exports map and I am happy with them so far.

**Comment by trusktr** [comment]
In my opinion, this should not be a GitHub discussion, but an issue. This is fairly critical, not a discussion topic for debate, and as more packages become Node ESM-flavored on npm, Meteor is going to lose compatibility with those packages.

At some point, the lack of support will be considered a bug, and it should be acted on in order to keep Meteor relevant.

**Comment by eric-burel** [comment]
Hi, you might want to take a look at "file-type", this is an example of a package that do not work with Meteor: https://www.npmjs.com/package/file-type

However to be completely fair, many packages may have weird exports and are not supported by many bundlers. For instance, @apollo/client doesn't load correctly in Next.js, and other Webpack 5-based bundlers, when it's a dependency of another ESM modules (like a package that exposes some hooks based on Apollo), because it doesn't have exports map.

**Comment by StorytellerCZ** [comment]
This just landed on Reify issue tracker, not sure about it, but might be useful: https://github.com/benjamn/reify/issues/365

**Comment by ajaypillay** [comment]
Hi, I've stumbled upon the ESM issue here:

https://github.com/orgs/mdx-js/discussions/2088
https://github.com/mdx-js/mdx/discussions/2085

I was wondering what the progress/stance is on whether Meteor is going to be supporting ESM packages, as I don't see any conclusion on what we should be doing moving forward. Thanks.

**Reply by StorytellerCZ** [reply]
Right now the Fibers issue takes priority over the others. I think we can expect movement on this after that.
On the Typescript side I think updating to 4.7 will be related to this as well from what I remember from the changelog.

**Comment by Grubba27** [comment]
As [commented](https://github.com/meteor/meteor/discussions/11727#discussioncomment-1553125) by @StorytellerCZ and requested by @zodern, I've updated the discussion name from:
 ESM packages support -> Feature: Support package.json exports field.
 If the name is still misleading, I can rename this until it makes sense :) 
Will update our roadmap with this discussion changes in a moment.

**Comment by jtarvainen** [comment]
Apologies if this is a silly question or the wrong place to ask, but should it be possible to create a Meteor app as an ESM? I've set up my project as an ESM, but this causes `meteor run` to crash with the following error:

> file:///my-app-path/.meteor/local/build/main.js:3
> process.chdir(require('path').join(__dirname, 'programs', 'server'));
>         ^
> 
> ReferenceError: require is not defined in ES module scope, you can use import instead
> This file is being treated as an ES module because it has a '.js' file extension and '/my-app-path/package.json' contains "type": "module". To treat it as a CommonJS script, rename it to use the '.cjs' file extension.
>     at file:///my-app-path/.meteor/local/build/main.js:3:9
>     at ModuleJob.run (internal/modules/esm/module_job.js:183:25)
>     at async Loader.import (internal/modules/esm/loader.js:178:24)
>     at async Object.loadESM (internal/process/esm_loader.js:68:5)
>     at async handleMainPromise (internal/modules/run_main.js:59:12)

Is there any way to avoid this?

**Reply by zodern** [reply]
Meteor transpiles ESM and internally uses `require` when it builds the app. Your app should run fine without having this error. It would be good to create a new issue for this bug.

**Reply by jtarvainen** [reply]
OK, thanks! I tried setting `"type": "module"` in my Meteor project's package.json, but that doesn't seem to work.

**Comment by brucejo75** [comment]
@Grubba27, @StorytellerCZ,  Should this issue be considered in v3.0?  Newer packages have started using it.

Thanks!

**Reply by StorytellerCZ** [reply]
Yes in the 3.x release line, not for 3.0 as that is already a big enough on its own and adding more onto it at this late stage will just delay it further.

**Reply by jankapunkt** [reply]
I think this should have been added way before 3.0 if Meteor finally wants to not only keep up but stay ahead of the majority of the broader ecosystem.

**Comment by domq** [comment]
I *think* I'm encountering this issue with Angular, see https://github.com/Urigo/angular-meteor/issues/1988

**Comment by StorytellerCZ** [comment]
Babel v8 is going to be ESM only: https://babeljs.io/docs/v8-migration

**Comment by MaxTwentythree** [comment]
Hi there,
is there any news to this?
I'm encountering this more and more and can't use some packages

**Reply by StorytellerCZ** [reply]
What version of Meteor are you on? Is this still issue in Meteor 3?

**Reply by MaxTwentythree** [reply]
Yeah, I'm on 3 (rc10).

**Comment by danopia** [comment]
I'm running into this issue on Meteor 3.0.1 when trying to update `octokit` which is a collection of NPM packages which import each other. The error messages are like `Error: Cannot find module '@octokit/request'` when one package tries importing another (I don't reference `@octokit/request` in my app).

Previously on Meteor 2.x I was stuck on an even older `octokit` from the old NodeJS version for Meteor 2's fibers. Now on Meteor 3, I've been able to update `octokit` a major version or two, but now further updates seem blocked on ESM. It's definitely progress 😅 I look forward to seeing this addressed in a future Meteor version.

**Comment by imajus** [comment]
Since this is a critical issue in my case which prevent me from creating new projects with Meteor, I've started investigation in attempt to fix the issue myself.

I don't know the internals of the Meteor Tool so I'll have to learn everything from scratch. This will be a whole new journey for me which make take several days so I've decided to record my daily progress here, for my own and everyone else's reference.

## Day №1

I've cloned the Meteor GitHub repository and compiled it locally. Now I can run local Meteor projects with a custom development version of Meteor Tool:

```
mymeteor run
```

With help of automatic debugging from JavaScript Debug Terminal in VSCode I can get my breakpoints work in the Meteor Tool source code when running any Meteor project locally.

Although, something is wrong with the file mappings so once debugger hits a breakpoint it opens a _runtime_ copy of the file in IDE as opposed to the original source file. 

![Screenshot from 2024-10-02 15-32-25](https://github.com/user-attachments/assets/7d3e8df5-5523-494f-9dad-cf910c7b77eb)

Maybe that applies only to TS files? Not sure yet, but that's a minor DX issue anyway.

Additionally, I've managed to do the following:
- Created a minimal example Meteor project with reproduction of the issue: https://github.com/imajus/meteor-exports-issue-example
- Found out the [file](https://github.com/meteor/meteor/blob/aac4b9441e2262e24cd7e68c7914abf43a418f3c/tools/isobuild/resolver.ts#L256) which most probably needs to be changed to fix the issue.
- Found the [logic](https://github.com/meteor/meteor/blob/aac4b9441e2262e24cd7e68c7914abf43a418f3c/tools/isobuild/resolver.ts#L256) which must be refactored first to not append extension automatically, because currently it maps `@insurgent/export-map-test/simple` to `@insurgent/export-map-test/simple.js` there.
- Found a [package](https://www.npmjs.com/package/resolve.exports) that I would most likely use in a `Resolver` to obtain an actual path based on `exports` property from the `package.json` file. The process of adding a new dependency to a Dev Bundle is tricky, but I managed to do it locally following the instruction from [DEVELOPMENT.md](https://github.com/meteor/meteor/blob/devel/DEVELOPMENT.md)

**Comment by imajus** [comment]
Alright, so for anyone following this...

## Day №2

- [Updated](https://github.com/imajus/meteor-exports-issue-example/commit/03ea66c6998162b2de0aedf17b19466e63a4f9a3) the example Meteor project by adding more imports
- Added `resolve.exports` Node package as a [dependency](https://github.com/meteor/meteor/commit/c9106375850ad024a639233374c2dd78c003f105) to the Dev Bundle
- Introduced a draft [fix](https://github.com/meteor/meteor/commit/0d76627941db61b406bf157e37dd61d2a92cd2c9) for errant logic appending `.js` file extension to the imported path which fixes this:
    > [...] currently it maps @insurgent/export-map-test/simple to @insurgent/export-map-test/simple.js there.

Now the import maps to a proper file path but the Bundler still fails with the same error. I'm assuming that my fix doesn't set some properties on the returned `resolved` object properly and the error still occurs deeper in the calls stack.

Before I continue, I'll have to spend more time acquainting myself with how the Meteor Bundler works internally.

**Reply by StorytellerCZ** [reply]
Amazing job so far! @nachocodoner might be able to provide insights.

**Comment by imajus** [comment]
## Day №3

I haven't had much time today to move the solution forward and instead concentrated on improving the DX. The goal is to make VSCode debugger work 100% properly and allow me to set breakpoints anywhere in Meteor source code.

Here's what I've got so far:

```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "type": "node",
      "request": "launch",
      "name": "Debug Meteor example",
      "runtimeExecutable": "${workspaceFolder}/dev_bundle/bin/node",
      "runtimeArgs": [
        "--max-old-space-size=4096",
        "--no-wasm-code-gc",
        "${workspaceFolder}/tools/index.js",
        "--exclude-archs=web.browser.legacy,web.cordova",
        "--raw-logs"
      ],
      "env": {
        "DEBUG": "*"
      },
      "cwd": "${workspaceFolder}/example",
      "outputCapture": "std",
      "restart": true
    }
  ]
}
```

This is intended to work with a project which resides in `examples` symlink inside the root of the project.

It allows to start the example Meteor application but doesn't seem to fix the source code mapping issue. Does anyone have any ideas on how to properly set that up?

**Reply by StorytellerCZ** [reply]
Side point here, maybe we should provide some default VS Code setup in the Meteor skeletons to make things easier. @nachocodoner @denihs

**Comment by imajus** [comment]
And, after some delay I'm back with ...

## Day №4

I've debugged more and realized that the problem now is that the Bundler does not create a symlink to the original package folder:

```
$ cd .meteor/local/build/programs/server/npm/node_modules
$ tree
.
├── @babel
│   └── runtime -> /home/denis/Workspaces/meteor/meteor-exports-issue-example/node_modules/@babel/runtime
├── @insurgent
├── meteor
│   ├── babel-compiler
│   │   └── node_modules
│   │       ├── acorn -> /home/denis/Workspaces/meteor/meteor-exports-issue-example/.meteor/local/isopacks/babel-compiler/npm/node_modules/acorn
│   │       ├── @ampproject
│   │       │   └── remapping -> /home/denis/Workspaces/meteor/meteor-exports-issue-example/.meteor/local/isopacks/babel-compiler/npm/node_modules/@ampproject/remapping
...
```

So the next step would be to find where these symlinks are generated and investigate why the package with `exports` property is not included.

**Comment by plbkbx** [comment]
Hello, do you have any updates on this? I’m encountering this issue more often and I'm unable to use certain packages. For example, the BPMN.io suite now only uses exports, making it very difficult to use.  This also affects the dynamic import of modules. Thanks.

**Comment by zodern** [comment]
I opened a couple PR's to implement support for resolving the exports field: https://github.com/meteor/meteor/pull/13520 and https://github.com/meteor/install/pull/1

**Comment by ceigey** [comment]
Thanks @zodern! Looking forward to it!

I was doing some experiments to see if you can include use Hono with `WebApp.handlers` (I'm embracing [serverlesslessness](https://youtu.be/aWfYxg-Ypm4?si=lyYLP-LcpGnxFDhe&t=347)), and realised the inbuilt middleware for Hono runs into this problem too as they've also adopted Node exports.

Importing directly from where the module really lives (e.g. `hono/dist/middleware/logger` instead of `hono/logger`) works, but then breaks TypeScript.

Realised you can use Babel & Vite aliases (both required at the same time if using [jorgenvatle:vite](https://packosphere.com/jorgenvatle/vite)) to work around the issue.

`vite.config.ts`:

```ts
export default defineConfig({
    // ...
    resolve: {
        alias: {
            'hono/logger': 'hono/dist/middleware/logger',
            'hono/request-id': 'hono/dist/middleware/request-id',
        }
    },
    // ...
})
```
`.babelrc` after installing `babel-plugin-module-resolver`:

```json
{
  "plugins": [
    [
      "module-resolver",
      {
        "root": ["."],
        "alias": {
          "hono/logger": "hono/dist/middleware/logger",
          "hono/request-id": "hono/dist/middleware/request-id"
        }
      }
    ]
  ],
  "presets": []
}

```

```diff
- import { logger } from 'hono/dist/middleware/logger'
- import { requestId } from 'hono/dist/middleware/request-id'
+ import { logger } from 'hono/logger'
+ import { requestId } from 'hono/request-id'
```

I'm assuming once this PR is in a stable release, things will probably break horribly with this config. But for now in my primitive test app I've got the imports working again and TypeScript isn't complaining (edit: and importantly the Hono handler works and is happily logging and generating request IDs)

_N.B.: I have not deployed this to Meteor cloud yet, I am only testing in local dev so far._

**Comment by banjerluke** [comment]
Thank you so much @ceigey for sharing. I found that Vite handles `exports` in `package.json` just fine out of the box, and everything worked fine with the dev server, but I'd get a "Cannot find module" error when running the built server bundle, e.g. with the [@sindresorhus/is](https://github.com/sindresorhus/is) package. Updating my `.babelrc` file as described in your post fixed this issue – didn't need to touch `vite.config.mts` in my case. Here's what I ended up adding to fix @sindresorhus/is imports:

```diff
  "plugins": [
    [
      "module-resolver",
      {
        "root": ["./"],
        "alias": {
          "@": "./imports",
+         "@sindresorhus/is": "@sindresorhus/is/distribution/index.js"
        }
      }
    ],
  ],
```

(Already had alias for `imports` directory, so that I could import from `"@/myfile"` instead of `"/imports/myfile"` or `"../../../myfile"`.)

Great to have this workaround until Meteor supports for the package.json exports field directly.

**Comment by aantich** [comment]
Hi, is there any update on the fix? Still having this issue with Swiper, React Router, React Markdown to name a few. Using latest Meteor (3.2.smth). Thank you!

**Reply by StorytellerCZ** [reply]
Here is the PR that works on this: https://github.com/meteor/meteor/pull/13520

**Reply by copleykj** [reply]
@aantich it's not the greatest DX, but a work around for this issue is to use a dependency injection pattern in your packages API. This moves the package import into the application code and out of the package build. I've recently used this pattern in `communitypackages:react-router-ssr` and while not quite the cleanest addition to the API, it works and it's not awful.
